File: crates/dewasm-backend-perl/units/wasi/fd_prestat_get.pl
# requires: memory/init
sub wasi_fd_prestat_get {
    my ($self, $fd, $out_ptr) = @_;
    my $e = $self->{fds}{$fd};
    return ERRNO_BADF unless defined($e) && $e->{dir} && defined($e->{preopen});
    # prestat: tag (u8, 0 = dir) + 3 pad + pr_name_len (u32).
    $self->{memory}->init($out_ptr, pack('CxxxV', 0, length($e->{preopen})), 0, 8);
    return ERRNO_SUCCESS;
}

File: crates/dewasm-backend-ruby/units/wasi/fd_prestat_get.rb
# requires: memory/init
def wasi_fd_prestat_get(fd, out_ptr)
  entry = @fds[fd]
  return ERRNO_BADF unless entry.is_a?(WasiDir) && entry.preopen_name
  # prestat: tag (u8, 0 = dir) + 3 pad + pr_name_len (u32)
  @memory.init(out_ptr, [0, entry.preopen_name.bytesize].pack("Cx3L<"), 0, 8)
  ERRNO_SUCCESS
end
